hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+ *
+ * @param[in]   callback    The callback function.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_set_callback_on_events(on_ble_evt_cb_t callback, void* context);
 
-// BLE_EVENT_CHAR_PARAM_CHANGED event data
-typedef struct ble_char_param_changed_event_data {
-    uint16_t conn_handle;
-    uint16_t char_handle;
-} ble_char_param_changed_event_data;
+/**
+ * Set local BLE identity address, which type must be either public or random.
+ *
+ * @param[in]   address Pointer to the BLE local identity address to be set.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_device_address(const hal_ble_addr_t* address);
 
-// BLE_EVENT_DATA_SENT event data
-typedef struct ble_data_sent_event_data {
-    uint16_t conn_handle;
-} ble_data_sent_event_data;
+/**
+ * Get local BLE identity address, which type is either public or random static.
+ *
+ * @param[out]  address Pointer to address structure to be filled in.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_device_address(hal_ble_addr_t* address);
 
-// BLE_EVENT_DATA_RECEIVED event data
-typedef struct ble_data_received_event_data {
-    uint16_t conn_handle;
-    uint16_t char_handle;
-    const char* data;
-    uint16_t size;
-} ble_data_received_event_data;
+/**
+ * Set the BLE device name.
+ *
+ * @param[in]   device_name Pointer to a UTF-8 encoded, <b>non NULL-terminated</b> string.
+ * @param[in]   len         Length of the UTF-8, <b>non NULL-terminated</b> string pointed to by device_name
+ *                          in octets (must be smaller or equal than @ref BLE_GAP_DEVNAME_MAX_LEN).
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_device_name(const uint8_t* device_name, uint16_t len);
 
-// Event handler callback
-typedef void(*ble_event_callback)(int event, const void* event_data, void* user_data);
+/**
+ * Get the BLE device name.
+ *
+ * @param[out]      device_name Pointer to an empty buffer where the UTF-8 <b>non NULL-terminated</b> string will be placed.
+ *                              Set to NULL to obtain the complete device name length.
+ * @param[in,out]   p_len       Length of the buffer pointed by device_name, complete device name length on output.
+ *
+ * @returns         0 on success, system_error_t on error.
+ */
+int ble_gap_get_device_name(uint8_t* device_name, uint16_t* len);
 
-// Profile
-typedef struct ble_profile {
-    uint16_t version; // API version
-    uint16_t reserved;
-    uint16_t flags;
-    uint16_t service_count;
-    ble_service* services;
-    const char* device_name;
-    const ble_manuf_data* manuf_data;
-    ble_event_callback callback;
-    void* user_data;
-} ble_profile;
+/**
+ * Set GAP Appearance value.
+ *
+ * @param[in]   appearance  Appearance (16-bit), see @ref BLE_APPEARANCES.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_appearance(uint16_t appearance);
```

Comment:
Consider using an enum for `appearance`, as it will provide better type safety.

Reply:
Using an integer type is safer from the ABI compatibility standpoint though

Reply:
If the enum values are not to be used as bitmasks, I'd still opt for enum for type safety. :man_shrugging: If we want to ensure ABI-compatibility we simply need to guarantee the size of the enum to not change no matter the compiler flags (e.g. `-fshort-enums`) or the platform. Enums are guaranteed to be not larger than 32-bit, so Vulkan for example does this simply by always adding `XXX_ENUM_MAX_VALUE = 0x7fffffff` to each enum.